qutebrowser/browser/downloads.py

```diff
@@ -894,110 +894,203 @@ def _on_data_changed(self, idx, *, webengine):
             qtutils.ensure_valid(end_index)
         self.dataChanged.emit(start_index, end_index)
 
-    def _raise_no_download(self, count):
-        """Raise an exception that the download doesn't exist.
+    def _select_downloads(self, count, all_, indexset, criterion, errmsg):
+        """Return selected downloads specified by indexset that satisfy the
+        criterion.
 
         Args:
-            count: The index of the download
+            count: count of download to get the actions applied against.
+            all_: whether all downloads should get the actions applied against.
+            indexset: Specify the set of download indexes to delete. E.g.
+                '2,5,6-last,3', where 'last' is special keyword that denotes
+                that latest download index. If you specify numbers smaller than
+                1, or greater than 'last', then those numbers will be silently
+                ignored.
+            criterion: A function that returns True only when satisfied.
+            errmsg: The postfix of the error message that is shown alongside
+                    list of downloads that do not match the criterion.
         """
-        if not count:
-            raise cmdexc.CommandError("There's no download!")
-        raise cmdexc.CommandError("There's no download {}!".format(count))
+        # test if there are any downloads at all
+        if len(self) == 0:
+            raise cmdexc.CommandError(
+                    "No downloads!".format(indexset)
+                  )
+
+        # support --all/-a
+        if all_ == True:
+            indexset = '1-last'
+            message.warning(
+                "'--all/-a' will be deprecated. Kindly use '0-last' instead."
+            )
+
+        # support count
+        if indexset is None:
+            if count == 0:
+                count = len(self)
+            indexset = str(count)
+
+        # parse the indexset into number intervals list
+        try:
+            indexset_p = utils.parse_numsettxt_into_numints(indexset, len(self))
+        except ValueError:
+            raise cmdexc.CommandError(
+                    "Invalid index set '{}'!".format(indexset)
+                  )
+
+        # identify downloads indexes that fall in the indexset
+        ins = utils.which_nums_in_numints(
```

Comment:
Yeah, that's what I suggest - making it its own class. I think giving the text in the constructor is nicer than having to set the `.text` attribute, so I'd prefer the first way. But you could also consider making a constructor that takes the "parsed" form (like `IndexSet((0, 3), (8, None))`), and the exposing a static method to parse a string (`IndexSet.from_string('0-3,8-last')`).

Depending on how you implement it, `IndexSet` or `NumberSet` is probably fine.

You could think about making it work with arbitrary `set`s (and `set` here means something that supports the `... in ...` check). This means that you could lazily build the union of any `set`s, and you could use the lazy `range` to make it work for indices:

```python
class SetUnion:
    def __init__(self, sets):
        self.sets = sets
    def __contains__(self, other):
        return any(other in s for s in self.sets)

indexset_p = SetUnion([range(0, 3), range(8, 10)])
2 in indexset_p # True
```

Now you have a pretty simple implementation that works for any sets. You lose some performance, as membership testing is now `O(number of sets)` (assuming that the inner sets support testing in constant time, which `range` does), but the code is very clear (and "number of sets" is probably within limits).